### PR TITLE
Fix #487: Reject case classes without parameter lists for Scala 2.11+

### LINKED ISSUE #487: Reject parsing of parameterless classes and case classes
Scalac:

``` scala
scala> case class A
<console>:1: error: case classes without a parameter list are not allowed;
use either case objects or case classes with an explicit `()' as a parameter list.
case class A
```

Currently, scala meta parses it correctly. I'm not sure if it's convenient to follow 2.11's scalac parser here. IIRC, older scala version parsed parameterless case classes correctly and only emitted warnings. Haven't checked, though.

BTW, the same happens for `implicit class A`, although we probably want to follow scalac's semantics here (currently, scalameta also parses it correctly).

``` scala
scala> implicit class A
<console>:13: error: implicit classes must accept exactly one primary constructor parameter
       implicit class A
```

### PR DESCRIPTION
**Pull request by jonas**

### CODE REVIEW

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 1]
```diff
@@ -3224,10 +3224,11 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
     }
     val className = typeName()
     rejectModCombination[Mod.Final, Mod.Sealed](mods, s"class $className")
-    // TODO:
-    // if (ofCaseClass && token.isNot[LeftParen])
-    //  syntaxError(token.offset, "case classes without a parameter list are not allowed;\n"+
-    //                             "use either case objects or case classes with an explicit `()' as a parameter list.")
+    if (mods.has[Mod.Case] && token.isNot[LeftParen] && !dialect.allowCaseClassWithoutParameterList) {
+      syntaxError("case classes without a parameter list are not allowed;\n"+
```
Comment (olafurpg) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: Can we use the same message as scalac?

> case classes must have a parameter list; try 'case class A()' or 'case object A'

File: scalameta/scalameta/shared/src/test/scala/scala/meta/tests/parsers/Scala211Suite.scala [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package scala.meta.tests
+package parsers
+
+import org.scalatest._
+import scala.meta._
+import scala.meta.dialects.Scala211
+
+class Scala211Suite extends ParseSuite with Matchers {
```
Comment (olafurpg) [inline scalameta/scalameta/shared/src/test/scala/scala/meta/tests/parsers/Scala211Suite.scala]: Let's use `assert` instead of matchers.

Comment (olafurpg) [inline scalameta/scalameta/shared/src/test/scala/scala/meta/tests/parsers/Scala211Suite.scala]: I struggle the parse the english-like `should have message "blah"` statements

File: scalameta/scalameta/shared/src/test/scala/scala/meta/tests/parsers/Scala211Suite.scala [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  test("case classes without a parameter list are not allowed") {
+    the[ParseException] thrownBy {
```
Comment (olafurpg) [inline scalameta/scalameta/shared/src/test/scala/scala/meta/tests/parsers/Scala211Suite.scala]:
```
val thrown = intercept[IndexOutOfBoundsException] {
 ...
}
assert(thrown.getMessage.contains("..."))
```

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 4]
```diff
@@ -3224,10 +3224,11 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
     }
     val className = typeName()
     rejectModCombination[Mod.Final, Mod.Sealed](mods, s"class $className")
-    // TODO:
-    // if (ofCaseClass && token.isNot[LeftParen])
-    //  syntaxError(token.offset, "case classes without a parameter list are not allowed;\n"+
-    //                             "use either case objects or case classes with an explicit `()' as a parameter list.")
+    if (mods.has[Mod.Case] && token.isNot[LeftParen] && !dialect.allowCaseClassWithoutParameterList) {
```
Comment (olafurpg) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: Let's start  this condition with `dialect.allow...`, that makes it easier to read.

File: scalameta/scalameta/shared/src/test/scala/scala/meta/tests/parsers/Scala211Suite.scala [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+      templStat("case class A")
```
Comment (olafurpg) [inline scalameta/scalameta/shared/src/test/scala/scala/meta/tests/parsers/Scala211Suite.scala]: What happens in the following cases?

```
scala> case class A @Inject() ()
defined class A

scala> case class A private ()
defined class A

scala> case class A private
<console>:1: error: case classes must have a parameter list; try 'case class A()' or 'case object A'
case class A private
                    ^

scala> case class A @Inject()
<console>:1: error: case classes must have a parameter list; try 'case class A()' or 'case object A'
case class A @Inject()
                      ^
scala> case class A[T]
<console>:1: error: case classes must have a parameter list; try 'case class A()' or 'case object A'
case class A[T]
               ^

scala> case class A[T] private
<console>:1: error: case classes must have a parameter list; try 'case class A()' or 'case object A'
case class A[T] private
                       ^
```

**Comment by olafurpg** [comment]
The validation seems to be a bit too strict at the moment https://travis-ci.org/scalameta/scalameta/jobs/240626757#L914

```
covariant *** FAILED ***
[info]   scala.meta.parsers.ParseException: <input>:1: error: case classes without a parameter list are not allowed;
[info] use either case objects or case classes with an explicit `()' as a parameter list.
[info] case class A[+T](t: T)
[info]             ^
[info]   at scala.meta.internal.parsers.Reporter$class.syntaxError(Reporter.scala:17)
[info]   at scala.meta.internal.parsers.Reporter$$anon$1.syntaxError(Reporter.scala:23)
[info]   at scala.meta.internal.parsers.Reporter$class.syntaxError(Reporter.scala:18)
[info]   at scala.meta.internal.parsers.Reporter$$anon$1.syntaxError(Reporter.scala:23)
[info]   at scala.meta.internal.parsers.ScalametaParser$$anonfun$classDef$1.apply(ScalametaParser.scala:3228)
[info]   at scala.meta.internal.parsers.ScalametaParser$$anonfun$classDef$1.apply(ScalametaParser.scala:3219)
[info]   at scala.meta.internal.parsers.ScalametaParser.atPos(ScalametaParser.scala:407)
[info]   at scala.meta.internal.parsers.ScalametaParser.classDef(ScalametaParser.scala:3219)
[info]   at scala.meta.internal.parsers.ScalametaParser.tmplDef(ScalametaParser.scala:3182)
[info]   at scala.meta.internal.parsers.ScalametaParser.defOrDclOrSecondaryCtor(ScalametaParser.scala:3037)
[info]   ...
```

**Comment by jonas** [comment]
@olafurpg Thanks for the review, do I get a medal for not doing anything right? ;)
I've updated the PR to address your comments and the check now works at the AST-level instead of at the token-level.

File: scalameta/scalameta/shared/src/test/scala/scala/meta/tests/parsers/Scala211Suite.scala (hunk 2 above)
Comment (xeno-by) [inline scalameta/scalameta/shared/src/test/scala/scala/meta/tests/parsers/Scala211Suite.scala]: Can we remove `Matchers` now that their functionality is unused?

File: scalameta/dialects/shared/src/main/scala/scala/meta/dialects/package.scala [hunk 6]
```diff
@@ -20,6 +20,10 @@ import scala.compat.Platform.EOL
   // Are extractor varargs specified using ats, i.e. is `case Extractor(xs @ _*)` legal or not?
   allowAtForExtractorVarargs: Boolean,
 
+  // Can case classes be declared without a parameter list?
+  // Deprecated in 2.10, not supported in 2.11 and newer.
+  allowCaseClassWithoutParameterList: Boolean,
```
Comment (xeno-by) [inline scalameta/dialects/shared/src/main/scala/scala/meta/dialects/package.scala]: Is this supported in Dotty? Let's have a test for this, too.

Comment (jonas) [inline scalameta/dialects/shared/src/main/scala/scala/meta/dialects/package.scala]: I get
```
15:11:10 ~ > dotc -repl
Welcome to Scala.next (pre-alpha, git-hash: 05aeeb5)  (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_73).
Type in expressions to have them evaluated.
Type :help for more information.
scala> case class A
-- [E004] Syntax Error: <console>:4:11 -----------------------------------------
4 |case class A
  |           ^
  |           A case class must have at least one parameter list

longer explanation available when compiling with `-explain`
```

**Comment by jonas** [comment]
Rebased with `Matchers` removed and tests for Dotty.

**Comment by xeno-by** [comment]
LGTM.

**Pull request merged**